core-server/src/main/java/org/glassfish/jersey/server/spi/ServerProvider.java

```diff
@@ -0,0 +1,90 @@
+ */
+
+package org.glassfish.jersey.server.spi;
+
+import java.net.URI;
+
+import javax.net.ssl.SSLContext;
+import javax.ws.rs.ConstrainedTo;
+import javax.ws.rs.ProcessingException;
+import javax.ws.rs.RuntimeType;
+
+import org.glassfish.jersey.server.ApplicationHandler;
+import org.glassfish.jersey.server.ResourceConfig;
+import org.glassfish.jersey.server.ServerFactory.SslClientAuth;
+import org.glassfish.jersey.spi.Contract;
+
+/**
+ * Service-provider interface for creating server instances.
+ *
+ * If supported by the provider, a server instance of the requested Java type
+ * will be created.
+ * <p>
+ * The created server is responsible for listening on a communication chanel for
+ * new client requests, dispatching these requests to the registered
+ * {@link ApplicationHandler Jersey application handler} using the handler's
+ * {@link ApplicationHandler#handle(org.glassfish.jersey.server.ContainerRequest)}
+ * handle(requestContext)} method and sending the responses provided by the
+ * application back to the client.
+ * </p>
+ * <p>
+ * A provider shall support a one-to-one mapping between a type, provided the
+ * type is not {@link Object}. A provider may also support mapping of sub-types
+ * of a type (provided the type is not {@code Object}). It is expected that each
+ * provider supports mapping for distinct set of types and subtypes so that
+ * different providers do not conflict with each other.
+ * </p>
+ * <p>
+ * An implementation can identify itself by placing a Java service provider
+ * configuration file (if not already present) -
+ * {@code org.glassfish.jersey.server.spi.ServerProvider} - in the resource
+ * directory {@code META-INF/services}, and adding the fully qualified
+ * service-provider-class of the implementation in the file.
+ * </p>
+ *
+ * @author Markus KARG (markus@headcrashing.eu)
+ *
+ */
+@Contract
+@ConstrainedTo(RuntimeType.SERVER)
+public interface ServerProvider {
+
+    /**
+     * Create a server of a given type which runs the given application defined by
+     * the given resource configuration.
+     *
+     * @param <T>
+     *            the type of the server.
+     * @param type
+     *            the type of the server.
+     * @param URI
+     *            uri the root address on which to bind the application.
+     * @param resourceConfig
+     *            The resource configuration defining the application.
+     * @param sslContext
+     *            The secure socket configuration to be used with HTTPS.
+     * @param sslClientAuth
+     *            Whether the server needs SSL client authentication.
+     * @return the server, otherwise {@code null} if the provider does not support
+     *         the requested {@code type}.
+     *
+     * @throws ProcessingException
+     *             if there is an error creating the server.
+     */
+    public <T extends Server> T createServer(Class<T> type, URI uri, ResourceConfig resourceConfig,
```

Comment:
This would be better  `createServer(Class<T> type, ResourceConfig resourceConfig, JAXRS.Configuration)`. The `Configuration` can contain all the necessary information for SSL, it even can contain the URI.
I would prefer `Application` over `ResourceConfig`. I know that the `*ServerFactory` classes work with `ResourceConfig`, but the `Container` classes are fine with `Application`, so a change on factories would be needed. This is could be done later.

Reply:
So it is ok that we directly use a JAX-RS interface inside of the Jersey backends? Good to know. :-)

Reply:
In this case, it would be aligned with ContainerProvider interface that is using the Application as well.

Reply:
I followed your proposals and pushed some additional commits reflecting them:
* `createServer` accepts `JAXRS.Configuration` instead of individual parameters.
* `ResourceConfig` was replaced by `Application`
* Factory methods have been added where needed.

Reply:
As you didn't respond for months, I assume you accept these changes, so I hereby mark this discussion as resolved. :-)